feat: Improve README for GitHub Marketplace DOCS-248 (#64)
codacy/codacy-coverage-reporter-action: README.md

[Codacy](https://www.codacy.com/) is an automated code review tool that makes it easy to ensure your team is writing high-quality code by analyzing more than 40 programming languages such as PHP, JavaScript, Python, Java, and Ruby. Codacy allows you to define your own quality rules, code patterns and quality settings you'd like to enforce to prevent issues on your codebase.

<br/>

<a href="https://www.codacy.com" target="_blank"><img src="images/codacy-logo.svg" alt="Codacy" width="400"/></a>

<br/>

## Uploading coverage to Codacy

Before setting up Codacy Coverage Reporter GitHub Action you must have tests and use tools to [generate coverage reports](https://docs.codacy.com/coverage-reporter/#generating-coverage) for the source code files in your repository.

The Codacy Coverage Reporter GitHub Action:

-   Uploads coverage report data to Codacy.
-   Prints the coverage report upload status on the console, which is visible on the workflow panel of the GitHub Action.
-   Fails the workflow if code coverage falls below your project’s [quality settings](https://docs.codacy.com/repositories-configure/adjusting-quality-settings/#gates).

> To upload coverage to Codacy using the GitHub Action with default settings:

Reviewer: Which are the "default settings"? Is there an option to use other settings?

Author: There are some extra configurations, but I have to admit that the "with default settings" bit adds little value (default settings should be...well, default), and may sound confusing. @prcr thoughts?

Reviewer 2: Indeed, the example use case is meant to instruct users to use the GitHub Action in the simplest (most straightforward) way possible, without going into details about extra or more advanced configurations.

So I do think that mentioning that in these instructions we'll be using the default settings is appropriate. :+1:

1.  Set up an API token to allow the GitHub Action to authenticate on Codacy:

    -   **If you're setting up one repository**, [obtain a project API token](https://docs.codacy.com/codacy-api/api-tokens/#project-api-tokens) and store it as an [encrypted secret for your **repository**](https://docs.github.com/actions/reference/encrypted-secrets#creating-encrypted-secrets-for-a-repository) with the name `CODACY_PROJECT_TOKEN`.
    -   **If you're setting up multiple repositories**, [obtain an account API token](https://docs.codacy.com/codacy-api/api-tokens/#account-api-tokens) and store it as an [encrypted secret for your **organization**](https://docs.github.com/en/actions/reference/encrypted-secrets#creating-encrypted-secrets-for-an-organization) with the name `CODACY_API_TOKEN`.

    > ⚠️ **Never write API tokens to your configuration files** and keep your API tokens well protected, as they grant owner permissions to your projects on Codacy.

2.  [Generate a supported code coverage report](https://docs.codacy.com/coverage-reporter/#generating-coverage) on each push to your repository.

> 3.  Add the following to a file `.github/workflows/codacy-coverage-reporter.yaml` in your repository. Update the sample code with the name of the environment storing the access token and the paths to the report files generated previously:

Reviewer: Having two sentences here makes it difficult to understand that "the following" refers to the sample.
Can we make this step similar to Step 3 in https://github.com/codacy/codacy-analysis-cli-action#integration-with-codacy-for-client-side-tools?
I'm not sure what we need to replace/update.

Author: Good idea. I'll also remove the reference to the access token, since it's `CODACY_PROJECT_TOKEN` and we discuss it in the point above.

    ```yaml
    name: Codacy Coverage Reporter

    on: ["push"]

    jobs:
      codacy-coverage-reporter:
        runs-on: ubuntu-latest
        name: codacy-coverage-reporter
        steps:
          - uses: actions/checkout@v2
          - name: Run codacy-coverage-reporter
            uses: codacy/codacy-coverage-reporter-action@v1
            with:
              project-token: ${{ secrets.CODACY_PROJECT_TOKEN }}
              # or
              # api-token: ${{ secrets.CODACY_API_TOKEN }}
              coverage-reports: <PATH_TO_REPORT>
              # or a comma-separated list for multiple reports
              # coverage-reports: <PATH_TO_REPORT>, <PATH_TO_REPORT>
    ```

4.  Optionally, to add coverage results to GitHub status checks, [configure a quality gate for coverage](https://docs.codacy.com/repositories-configure/integrations/github-integration/#status-checks) and [enable sending status checks](https://docs.codacy.com/repositories-configure/integrations/github-integration/#status-checks) on Codacy.

## Extra configurations

> The Codacy GitHub Action is a wrapper for running the [Codacy Coverage Reporter CLI](https://github.com/codacy/codacy-coverage-reporter#codacy-coverage-reporter) and supports a subset of the parameters available for the command `report`. In addition to `coverage-reports`, `project-token`, and `api-token`, described in the example above, you can set:

Author: I'm not super happy with the "in addition to" transition here. It's as if these parameters have intrinsically different weight/importance. Should we harmonize and just reserve a line for each?

Reviewer 2: Yes, that makes perfect sense. I agree with you that it's best to list all supported parameters at the same level (on the same list).

> * `language`: optionally force associating a language with your coverage report(s).
> * `force-coverage-parser`: optionally force using a specific coverage report parser.

Reviewer 2: Since these options aren't mandatory, it already becomes implicit that they're optional. If we don't mention that, the descriptions become simpler (and it also becomes easier to write the descriptions for the `project-token` and `api-token`, which are mutually exclusive :slightly_smiling_face:).

A small stylistic detail regarding lists is that [we can normally do without the periods](https://learn.microsoft.com/en-us/style-guide/scannable-content/lists#punctuation) at the end of each list item (it's not wrong having the periods, but I do find this style more minimalist).

```suggestion
* `language`: force associating a language with your coverage report(s)
* `force-coverage-parser`: force using a specific coverage report parser
```

I checked the corresponding section for the [Codacy Analysis CLI GitHub Action](https://github.com/codacy/codacy-analysis-cli-action#extra-configurations) and I like the syntax that you used for describing each option here better. When you're tweaking the other section, you should apply the same syntax there as well. :nerd_face:

    -   **If you're setting up multiple repositories**, [obtain an account API token](https://docs.codacy.com/codacy-api/api-tokens/#account-api-tokens) and store it as an [encrypted secret for your **organization**](https://docs.github.com/en/actions/reference/encrypted-secrets#creating-encrypted-secrets-for-an-organization) with the name `CODACY_API_TOKEN`.

    > ⚠️ **Never write API tokens to your configuration files** and keep your API tokens well protected, as they grant owner permissions to your projects on Codacy.

2.  [Generate a supported code coverage report](https://docs.codacy.com/coverage-reporter/#adding-coverage-to-your-repository) before each push to your repository.

   3.  Add the following to a file `.github/workflows/codacy-coverage-reporter.yaml` in your repository. Update the sample code with the name of the environment storing the access token and the paths to the report files generated previously:

       ```yaml
       name: Codacy Coverage Reporter

       on: ["push"]

       jobs:
         codacy-coverage-reporter:
           runs-on: ubuntu-latest
           name: codacy-coverage-reporter

>            environment: production # Environment storing the access token

Author: This is more of a GitHub-related step, but I had to add it to get it to work properly on my tests. Worth mentioning IMO.

Author: We may also want to mention this in the Analysis CLI Action repo.

           steps:
             - uses: actions/checkout@v2
             - name: Run codacy-coverage-reporter
               uses: codacy/codacy-coverage-reporter-action@v1
               with:
                 project-token: ${{ secrets.CODACY_PROJECT_TOKEN }}
                 # or
                 # api-token: ${{ secrets.CODACY_API_TOKEN }}
                 coverage-reports: path/to/coverage.xml
                 # or a comma-separated list for multiple reports
                 # coverage-reports: package1/coverage.xml, package2/coverage.xml

>                  # language: '' # Optionally force associating a language with your coverage report

Author: Haven't tested this. Do we want to include it? I can look into it.

>                  # force-coverage-parser: '' # Optionally force using a specific coverage report parser

Author: Same as above for `language`

> 4.  Optionally, to have information about the coverage results directly on your pull requests, [enable the GitHub integration](https://docs.codacy.com/repositories-configure/integrations/github-integration/) on Codacy.

Author: I reworded the original, from:
"Do *action* to *achieve goal*."
to:
"To *achieve goal*, do *action*."

Seems more sensible to me. Thoughts?

[![Codacy Badge](https://app.codacy.com/project/badge/Grade/db5eda4c50b24c008b2c19f1117c9cef)](https://www.codacy.com/gh/codacy/codacy-coverage-reporter-action/dashboard?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=codacy/codacy-coverage-reporter-action&amp;utm_campaign=Badge_Grade)

> GitHub Action for [uploading coverage reports to Codacy](https://docs.codacy.com/coverage-reporter/#uploading-coverage) on all commits and pull requests, tracking code coverage, and returning code coverage issues.

Reviewer 2: As we say, the Codacy Coverage Reporter (and subsequently, the GitHub Action) acts just as an uploader of the coverage reports to Codacy. This part should still mention the added benefits of uploading coverage to Codacy, but it must be rewritten. It could be something like this:

```suggestion
GitHub Action for [uploading coverage reports to Codacy](https://docs.codacy.com/coverage-reporter/#uploading-coverage) on all commits and pull requests, allowing you to track code coverage on your repository and to flag pull requests that cause coverage to drop or add code that isn't tested.
```

[Codacy](https://www.codacy.com/) is an automated code review tool that makes it easy to ensure your team is writing high-quality code by analyzing more than 40 programming languages such as PHP, JavaScript, Python, Java, and Ruby. Codacy allows you to define your own quality rules, code patterns and quality settings you'd like to enforce to prevent issues on your codebase.

<br/>

<a href="https://www.codacy.com" target="_blank"><img src="images/codacy-logo.svg" alt="Codacy" width="400"/></a>

<br/>

> ## Coverage using your project’s quality settings

Reviewer 2: Here, we can make the title more straightforward (even if it loses some consistency with the ones from the Codacy Analysis CLI):

```suggestion
## Uploading coverage to Codacy
```

In practice, the GitHub Action knows nothing about the repository quality settings configured on Codacy. The use case for the GitHub Action is just to ensure that it uploads the relevant coverage reports for each commit pushed to the repository.

Before setting up Codacy Coverage Reporter GitHub Action you must have tests and use tools to [generate coverage reports](https://docs.codacy.com/coverage-reporter/#generating-coverage) for the source code files in your repository.

The Codacy Coverage Reporter GitHub Action:

-   Uploads coverage report data to Codacy.
-   Prints the coverage report upload status on the console, which is visible on the workflow panel of the GitHub Action.

> -   Fails the workflow if code coverage falls below your project’s [quality settings](https://docs.codacy.com/repositories-configure/adjusting-quality-settings/#gates).

Reviewer 2: I propose that we simplify things a bit and we actually remove these items:

 - "Prints the coverage report upload status on the console" :arrow_right: this is already somewhat to be expected, as most GitHub actions do output some kind of status or even debugging information to the console
 - "Fails the workflow if code coverage" :arrow_right: the GitHub action uploads the coverage data but doesn't get back any results from Codacy. It's Codacy itself that may post a failed status check if configured to do so.

> To use the GitHub Action with default settings:

Reviewer 2: By removing the list items above, we're left with just one item ("Uploads coverage report data to Codacy") so it doesn't make sense to have a list anymore. But we can still make it explicit that the use case is for uploading coverage to Codacy in the sentence that introduces the procedure:

```suggestion
To upload coverage to Codacy using the GitHub Action with default settings:
```

1.  Set up an API token to allow the GitHub Action to authenticate on Codacy:

    -   **If you're setting up one repository**, [obtain a project API token](https://docs.codacy.com/codacy-api/api-tokens/#project-api-tokens) and store it as an [encrypted secret for your **repository**](https://docs.github.com/actions/reference/encrypted-secrets#creating-encrypted-secrets-for-a-repository) with the name `CODACY_PROJECT_TOKEN`.
    -   **If you're setting up multiple repositories**, [obtain an account API token](https://docs.codacy.com/codacy-api/api-tokens/#account-api-tokens) and store it as an [encrypted secret for your **organization**](https://docs.github.com/en/actions/reference/encrypted-secrets#creating-encrypted-secrets-for-an-organization) with the name `CODACY_API_TOKEN`.

    > ⚠️ **Never write API tokens to your configuration files** and keep your API tokens well protected, as they grant owner permissions to your projects on Codacy.

> 2.  [Generate a supported code coverage report](https://docs.codacy.com/coverage-reporter/#adding-coverage-to-your-repository) before each push to your repository.

Reviewer 2: * I think it's more appropriate to link directly to the section on how to generate the coverage reports (just like in the first sentence of this section).
* The tests that generate the coverage reports should run in the CI/CD pipeline that's triggered by each push to the repository. So, it's more correct to say that the user must generate the coverage reports "on each push" instead of "before each push" to the repository.

```suggestion
2.  [Generate a supported code coverage report](https://docs.codacy.com/coverage-reporter/#generating-coverage) on each push to your repository.
```

3.  Add the following to a file `.github/workflows/codacy-coverage-reporter.yaml` in your repository. Update the sample code with the name of the environment storing the access token and the paths to the report files generated previously:

    ```yaml
    name: Codacy Coverage Reporter

    on: ["push"]

    jobs:
      codacy-coverage-reporter:
        runs-on: ubuntu-latest
        name: codacy-coverage-reporter

>         environment: production # Environment storing the access token

Reviewer 2: It's not mandatory to associate GitHub actions workflows with environments, as [this is an optional feature](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment). :thinking: So, my suggestion is to simplify the example and not mention this here:

```suggestion
```

I believe that you ran into this because in your tests you configured the secret `CODACY_PROJECT_TOKEN` under a ["sandbox" environment](https://github.com/codacy-docs/sandbox/settings/environments/729300222/edit). But you could have created a repository secret [directly here](https://github.com/codacy-docs/sandbox/settings/secrets/actions) without having to be associated with an environment. :slightly_smiling_face:

        steps:
          - uses: actions/checkout@v2
          - name: Run codacy-coverage-reporter
            uses: codacy/codacy-coverage-reporter-action@v1
            with:
              project-token: ${{ secrets.CODACY_PROJECT_TOKEN }}
              # or
              # api-token: ${{ secrets.CODACY_API_TOKEN }}
              coverage-reports: path/to/coverage.xml
              # or a comma-separated list for multiple reports
              # coverage-reports: package1/coverage.xml, package2/coverage.xml

>               # language: '' # Optionally force associating a language with your coverage report
>               # force-coverage-parser: '' # Optionally force using a specific coverage report parser

Reviewer 2: As we realized earlier today, it's better to mention the extra settings that the GitHub Action allows to specify in a separate section at the end.

This allows us to have a simpler code example focused on the main use case.

```suggestion
```

    ```

> 4.  Optionally, to have information about the coverage results directly on your pull requests, [enable the GitHub integration](https://docs.codacy.com/repositories-configure/integrations/github-integration/) on Codacy.

Reviewer 2: Currently, the only information about coverage that Codacy posts on pull requests are the **coverage status checks** that are mentioned in the last step (the users must configure a coverage quality gate and enable the status checks).

So, this step is actually redundant. :slightly_smiling_face: 

```suggestion
```

              # coverage-reports: <PATH_TO_REPORT>, <PATH_TO_REPORT>
    ```

4.  Optionally, to add coverage results to GitHub status checks, [configure a quality gate for coverage](https://docs.codacy.com/repositories-configure/integrations/github-integration/#status-checks) and [enable sending status checks](https://docs.codacy.com/repositories-configure/integrations/github-integration/#status-checks) on Codacy.

## Options

The Codacy GitHub Action is a wrapper for running the [Codacy Coverage Reporter CLI](https://github.com/codacy/codacy-coverage-reporter#codacy-coverage-reporter) and supports a subset of the options available for the command `report`:

- `coverage-reports`: path to the coverage report relative to the repository root, or a comma-separated list for multiple reports
- `project-token`: [project API token](https://docs.codacy.com/codacy-api/api-tokens/#project-api-tokens)
- `api-token`: [account API token](https://docs.codacy.com/codacy-api/api-tokens/#account-api-tokens), an alternative to `project-token` when setting up multiple repositories
- `language` (optional): associate a language with your coverage report(s)

> - `force-coverage-parser` (optional): force using a specific coverage report parser

Author: By listing the options all together, I was forced to make some choices. Let me know if they make sense:

- The title "extra configurations" now should be called "options" or similar, as we're listing all of them.
  - On a related note, these were called "parameters" but I've seen "options" used more often for CLI commands. I'm open to reverting to "parameters".
- I had to reinstate the "optional" flag for `language` and `force-coverage-parser` to differentiate them, since they're not the only ones anymore.
- The last two descriptions are imperative sentences, the first three are sentence fragments (I believe) which makes the list [nonparallel](https://developers.google.com/tech-writing/one/lists-and-tables#exercise_1). Do we care?

Reviewer 2: Now that I'm also having a look at the full list of options, I realize that they're easier to scan and read when formatted in a table like [what we already had before](https://github.com/codacy/codacy-coverage-reporter-action#workflow-options)!

So, my suggestion now is that you keep exactly the same text that you have (which is an improvement over what we currently have live) but use a table to format the information.

> * The title "extra configurations" now should be called "options" or similar, as we're listing all of them.

Makes sense! :+1: 
 
>   * On a related note, these were called "parameters" but I've seen "options" used more often for CLI commands. I'm open to reverting to "parameters".

Also sounds good. :+1: 

> * I had to reinstate the "optional" flag for `language` and `force-coverage-parser` to differentiate them, since they're not the only ones anymore.

Also, notice that the option `coverage-reports` [is also optional](https://github.com/codacy/codacy-coverage-reporter-action/blob/master/action.yml#L14-L17) - if the user doesn't specify the reports, the Codacy Coverage Reporter will try to detect and upload [these supported coverage reports](https://docs.codacy.com/coverage-reporter/#generating-coverage) from the root of the repository.

In fact, all options are optional. But you do need either a project or an account API token for authenticating to Codacy.

> * The last two descriptions are imperative sentences, the first three are sentence fragments (I believe) which makes the list [nonparallel](https://developers.google.com/tech-writing/one/lists-and-tables#exercise_1). Do we care?

It's not a big deal, and you could adjust the last two descriptions to read like this, for example:

"specifies a language to associate with your coverage report"
"forces using a specific coverage report parser"

But the imperative actually seems more straightforward to me.
